### ISSUE TITLE
Missing Support for Vulkan RAII Headers

### ISSUE THREAD

**Issue by jnhyatt** [YaaZ/VulkanMemoryAllocator-Hpp#15]
Alongside Vulkan-Hpp, the library has included support for RAII types, and is one of the cleanest ways of writing Vulkan code. As VulkanMemoryAllocator follows Vulkan's design philosophy so closely, and as VulkanMemoryAllocator-Hpp also mirrors Vulkan-Hpp, I think it would be beneficial to add support for RAII types in this project as well.

To address the obvious question of "why not just use `UniqueHandle`", I'd use the same reasoning behind why many prefer `vk::raii` to `UniqueHandle`. The RAII handle wrappers do more than simply properly destroy the object when it goes out of scope -- it also ensures that *no invalid handle can exist*. It is perfectly valid (just as with `std::unique_ptr`) to default construct a `UniqueHandle`, resulting in an invalid object that must be queried for validity before it can be used. The RAII wrappers generally result in much more expressive code and lower potential for bugs to be introduced:

```c++
vk::raii::Device device(physicalDevice, vk::DeviceCreateInfo(...));

```

Or for VMA RAII:

```c++
vma::raii::Allocator allocator(vma::AllocatorCreateInfo(...));
```
`vk::raii::Device` has no default constructor, and all constructors throw on error, meaning no invalid `vk::raii::Device` can be created. RAII handles also define a move constructor and delete the copy constructor to ensure each handle is unique.

Additionally, the RAII handles provide helper methods to move functions into handles they are most closely related to. Instead of:

```c++
device.bindBufferMemory(buffer, memory, memoryOffset);
void* pBuffer = allocator.mapMemory(allocation);
```

We have the following:

```c++
buffer.bindMemory(*memory, memoryOffset);
void* pBuffer = allocation.mapMemory();
```

The syntax is cleaner, and more importantly, the semantics are expressed clearly. The number of arguments per function call is reduced, and functionality is moved into the most strongly related handle class. Documentation is easier to browse and understand, and the code is generally more readable. The generator rules would likely need to be modified to support the new syntax, and `vma::raii::Allocator` would need to take a `vk::raii::Context` in its constructor, but it would certainly add consistency across the Vulkan ecosystem for anyone using the RAII library. `<vulkan_raii.hpp>` is a valuable library that I believe would make a great addition to VulkanMemoryAllocator-Hpp.

For reference, here is the [Vulkan RAII Programming Guide](https://github.com/KhronosGroup/Vulkan-Hpp/blob/master/vk_raii_ProgrammingGuide.md), it includes a good description of all the benefits using the RAII headers can provide.

**Comment by YaaZ** [comment]
Agree, that would be very nice to have. And there's also one cool thing about RAII you didn't mention - each object keeps its dispatcher instead of it being a template argument, so there's no template boilerplate or implicit global state (`VULKAN_HPP_DEFAULT_DISPATCHER`) as with `UniqueHandle`.

The problem with generating RAII is that there are much more exceptions than with unique handles, so it will end up like 50% generated and 50% handwritten. I hope some time I'll find a time to implement this if nobody does this first :)

And regarding passing `vk::raii::Context` to `vma::raii::Allocator` constructor, there's already an utility function which can extract `vma::VulkanFunctions` from RAII dispatcher: https://github.com/YaaZ/VulkanMemoryAllocator-Hpp/issues/11.

**Comment by jnhyatt** [comment]
I'm happy to get started on it, but I thought I'd make sure there wasn't any work already done first. I've been poking through the generator code and it doesn't seem too bad. I'll open a branch in my fork and put a link here once I've got something to show

**Comment by YaaZ** [comment]
Cool, thank you!

**Comment by rickyjames35** [comment]
I'm looking forward to this feature. Any plans to still implement it?

**Comment by miyoku157** [comment]
I'm also looking for this feature to be implemented. But meanwhile, is there a way to make vma-hpp interact with vulkan-hpp raii? Or we Just can't use them at All?

**Comment by AbitTheGray** [comment]
Hello, anyone still working on this?

If not, I may give it a try myself as `vk::raii` helps me a lot but combining it with non-RAII `vma` pollutes the code (compared to RAII variant). Is there at least a prototype?

**Comment by YaaZ** [comment]
Sure, would be great if you could take this. There's nobody working on it AFAIK. 
I found my old snippet where I tried playing with hand-written RAII, it's dirty and won't compile right away, but here you go anyway: https://pastebin.com/LKATTafT

**Comment by AbitTheGray** [comment]
I gave it a try myself - [vk_mem_alloc_raii.hpp](https://gist.github.com/AbitTheGray/78a25a3b5b316b3840bf2f7c7d619638)
Handwritten C++ code based on `vulkan/vulkan_raii.hpp` and functions from `vma::Allocator`. I did not have time to test it (and don't know when I'll have time for it again, hopefully soon) - use at your own risk, preferably not in production.

I've included the same license in file's header - CC0 1.0 Universal.

**Comment by dsvensson** [comment]
@AbitTheGray @YaaZ Could you perhaps take this over the finishing line with some feedback on @AbitTheGray's solution and turn it into a PR?

**Comment by sharadhr** [comment]
Okay, it looks like @YaaZ has straight up written an entire mini C++ parser in the [new branch](/YaaZ/VulkanMemoryAllocator-Hpp/tree/generator-dev), and I am far too stupid to be able to parse (heh) this in one go to submit a PR. Hence, I'd like to ask @YaaZ to keep in mind the latest changes to Vulkan-Hpp and the module, and adopt a similar strategy of 'all-in-on-modules' or no modules.

**Comment by YaaZ** [comment]
Now that CI is set up and green, all known issues addressed, I'm waiting for https://github.com/KhronosGroup/Vulkan-Hpp/pull/2349 and https://github.com/KhronosGroup/Vulkan-Hpp/pull/2357 and then gonna make a release. If you haven't found any issues in the [dev branch](https://github.com/YaaZ/VulkanMemoryAllocator-Hpp/tree/generator-dev) yet, it's time!
